proxy/src/proxy/conntrack.rs

```diff
@@ -0,0 +1,680 @@
+    #[default]
+    Init = 0,
+    Idle = 1,
+    Transaction = 2,
+    Busy = 3,
+    Closed = 4,
+    Unknown = 5,
+}
+
+impl ConnectionState {
+    const fn into_repr(self) -> u8 {
+        self as u8
+    }
+
+    const fn from_repr(value: u8) -> Option<Self> {
+        Some(match value {
+            0 => Self::Init,
+            1 => Self::Idle,
+            2 => Self::Transaction,
+            3 => Self::Busy,
+            4 => Self::Closed,
+            5 => Self::Unknown,
+            _ => return None,
+        })
+    }
+}
+
+impl fmt::Display for ConnectionState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match *self {
+            ConnectionState::Init => f.write_str("init"),
+            ConnectionState::Idle => f.write_str("idle"),
+            ConnectionState::Transaction => f.write_str("transaction"),
+            ConnectionState::Busy => f.write_str("busy"),
+            ConnectionState::Closed => f.write_str("closed"),
+            ConnectionState::Unknown => f.write_str("unknown"),
+        }
+    }
+}
+
+/// Stores the `ConnectionState`. Used by ConnectionTracker to avoid needing
+/// mutable references.
+#[derive(Debug, Default)]
+struct AtomicConnectionState(AtomicU8);
+
+impl AtomicConnectionState {
+    fn set(&self, state: ConnectionState) {
+        self.0.store(state.into_repr(), Ordering::Relaxed);
+    }
+
+    fn get(&self) -> ConnectionState {
+        ConnectionState::from_repr(self.0.load(Ordering::Relaxed)).expect("only valid variants")
+    }
+}
+
+/// Tracks the `ConnectionState` of a connection by inspecting the frontend and
+/// backend stream and reacting to specific messages. Used in combination with
+/// two `TrackedStream`s.
+pub struct ConnectionTracker<SCO: StateChangeObserver> {
+    state: AtomicConnectionState,
+    observer: SCO,
+    conn_id: SCO::ConnId,
+}
+
+impl<SCO: StateChangeObserver> Drop for ConnectionTracker<SCO> {
+    fn drop(&mut self) {
+        self.observer
+            .change(self.conn_id, self.state.get(), ConnectionState::Closed);
+    }
+}
+
+impl<SCO: StateChangeObserver> ConnectionTracker<SCO> {
+    pub fn new(conn_id: SCO::ConnId, observer: SCO) -> Self {
+        ConnectionTracker {
+            conn_id,
+            state: AtomicConnectionState::default(),
+            observer,
+        }
+    }
+
+    pub fn frontend_message_tag(&self, tag: Tag) {
+        self.update_state(|old_state| Self::state_from_frontend_tag(old_state, tag));
+    }
+
+    pub fn backend_message_tag(&self, tag: Tag) {
+        self.update_state(|old_state| Self::state_from_backend_tag(old_state, tag));
+    }
+
+    fn update_state(&self, new_state_fn: impl FnOnce(ConnectionState) -> ConnectionState) {
+        let old_state = self.state.get();
+        let new_state = new_state_fn(old_state);
+        if old_state != new_state {
+            self.observer.change(self.conn_id, old_state, new_state);
+            self.state.set(new_state);
+        }
+    }
```

Comment:
This is effectively a SPMC setup, right? One connection broadcasts event changes.

I wonder if it would be better to make this function `&mut self` as the single producer, and to cache the current state to avoid the atomic load.

I'm especially worried due to the relaxed atomics - I'm not sure if we can rely that a read-after-write is consistent if the tokio task moves to another thread.

Reply:
Right, the atomic ops need to be ordered. Either Acq+Rel or even use CAS for the set(). We don't have the risk of ABA here.

I'll think about `&mut`. The idea of the atomic was to avoid that, but we can wrap put the ConnTracker into `Arc<Mutex<<..>>`.

Reply:
My understanding is that only the observer is shared, but the tracker is owned by the stream, right?

Reply:
> We don't have the risk of ABA here.

Actually, with CAS we do. We can do with ArcMutex for now and optimize later.

Reply:
> The ConnTracker is per conn and shared between the stream pairs

I wonder if it could be on just the one stream object, with one side processing on `poll_read()`, and the other on `poll_write()` 🤔

Reply:
Sure, that's possible. Then the ConnTracker has both state machines, one for each direction, but I think we still need synchronization around the observer. I will try a few things.